examples/federation-axum/instance.rs

```diff
@@ -0,0 +1,163 @@
+    /// This holds all library data
+    local_instance: LocalInstance,
+    /// Our "database" which contains all known users (local and federated)
+    pub users: Mutex<Vec<MyUser>>,
+    /// Same, but for posts
+    pub posts: Mutex<Vec<MyPost>>,
+}
+
+/// Use this to store your federation blocklist, or a database connection needed to retrieve it.
+#[derive(Clone)]
+struct MyUrlVerifier();
+
+#[async_trait]
+impl UrlVerifier for MyUrlVerifier {
+    async fn verify(&self, url: &Url) -> Result<(), &'static str> {
+        if url.domain() == Some("malicious.com") {
+            Err("malicious domain")
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl Instance {
+    pub fn new(hostname: String) -> Result<InstanceHandle, Error> {
+        let settings = InstanceSettings::builder()
+            .debug(true)
+            .url_verifier(Box::new(MyUrlVerifier()))
+            .build()?;
+        let local_instance =
+            LocalInstance::new(hostname.clone(), Client::default().into(), settings);
+        let local_user = MyUser::new(generate_object_id(&hostname)?, generate_actor_keypair()?);
+        let instance = Arc::new(Instance {
+            local_instance,
+            users: Mutex::new(vec![local_user]),
+            posts: Mutex::new(vec![]),
+        });
+        Ok(instance)
+    }
+
+    pub fn local_user(&self) -> MyUser {
+        self.users.lock().unwrap().first().cloned().unwrap()
+    }
+
+    pub fn local_instance(&self) -> &LocalInstance {
+        &self.local_instance
+    }
+
+    pub fn listen(instance: &InstanceHandle) -> Result<(), Error> {
+        let hostname = instance.local_instance.hostname();
+        let instance = instance.clone();
+        let app = Router::new()
+            .route("/objects/:user_name", get(http_get_user))
+            .layer(
+                ServiceBuilder::new()
+                    .map_request_body(body::boxed)
+                    .layer(middleware::from_fn(verify_request_payload)),
+            )
+            .with_state(instance);
+
+        // run it
+        let addr = SocketAddr::from_str(hostname)?;
+        let server = axum::Server::bind(&addr).serve(app.into_make_service());
+
+        task::spawn(server);
+        Ok(())
+    }
+}
+
+/// FIXME
+use axum_macros::debug_handler;
+#[debug_handler(body = Body)]
+/// Handles requests to fetch user json over HTTP
+async fn http_get_user(
+    State(data): State<InstanceHandle>,
+    request: Request<Body>,
+) -> impl IntoResponse {
+    let hostname: String = data.local_instance.hostname().to_string();
+    let request_url = format!("http://{}{}", hostname, &request.uri());
+
+    let url = Url::parse(&request_url).expect("Failed to parse url");
+
+    let user = ObjectId::<MyUser>::new(url)
+        .dereference_local(&data)
+        .await
+        .expect("Failed to dereference user")
+        .into_apub(&data)
+        .await
+        .expect("Failed to convert to apub user");
+
+    let json =
```

Comment:
Not sure if more or less elegant but I did add myself this small wrapper for doing this:

```
#[derive(Debug, Clone, Copy, Default)]
struct ApubJson<Json: Serialize>(Json);

impl<Json: Serialize> IntoResponse for ApubJson<Json> {
    fn into_response(self) -> axum::response::Response {
        let mut response = axum::response::Json(self.0).into_response();
        response.headers_mut().insert(
            header::CONTENT_TYPE,
            APUB_JSON_CONTENT_TYPE.parse().unwrap(),
        );
        response
    }
}
```

The benefit is that instead of `impl IntoResponse` I can return `Result<ApubJson<Person>, ServerError>` which makes it possible to use typed errors :)

Maybe thats something that could be upstreamed into this PR?